packages/@aws-cdk/aws-stepfunctions-tasks/lib/sagemaker/create-model.ts

```diff
@@ -0,0 +1,258 @@
+                }),
+              ],
+            }),
+          },
+        });
+
+    // create a security group if not defined
+    if (this.vpc && this.securityGroup === undefined) {
+      this.securityGroup = new ec2.SecurityGroup(this, 'ModelSecurityGroup', {
+        vpc: this.vpc,
+      });
+      this.connections.addSecurityGroup(this.securityGroup);
+      this.securityGroups.push(this.securityGroup);
+    }
+    const stack = cdk.Stack.of(this);
+    // https://docs.aws.amazon.com/sagemaker/latest/dg/api-permissions-reference.html
+    return [
+      new iam.PolicyStatement({
+        actions: ['sagemaker:CreateModel'],
+        resources: [
+          stack.formatArn({
+            service: 'sagemaker',
+            resource: 'model',
+            // If the model name comes from input, we cannot target the policy to a particular ARN prefix reliably.
+            // SageMaker uses lowercase for resource name in the arn
+            resourceName: sfn.JsonPath.isEncodedJsonPath(this.props.modelName) ? '*' : `${this.props.modelName.toLowerCase()}*`,
+          }),
+        ],
+      }),
+      new iam.PolicyStatement({
+        actions: ['sagemaker:ListTags'],
+        resources: ['*'],
+      }),
+      new iam.PolicyStatement({
+        actions: ['iam:PassRole'],
+        resources: [this._role!.roleArn],
+        conditions: {
+          StringEquals: { 'iam:PassedToService': 'sagemaker.amazonaws.com' },
+        },
+      }),
+    ];
+  }
+
+  private renderTags(tags: { [key: string]: any } | undefined): { [key: string]: any } {
+    return tags ? { Tags: Object.keys(tags).map((key) => ({ Key: key, Value: tags[key] })) } : {};
+  }
+
+  private renderVpcConfig(config: VpcConfig | undefined): { [key: string]: any } {
+    return config
+      ? {
+        VpcConfig: {
+          SecurityGroupIds: cdk.Lazy.listValue({ produce: () => this.securityGroups.map((sg) => sg.securityGroupId) }),
+          Subnets: this.subnets,
+        },
+      }
+      : {};
+  }
+
+  private renderPrimaryContainer(config: ContainerDefinition): {[key: string]: any} {
+    return (config) ? { PrimaryContainer: this.renderContainer(config) } : {};
+  }
+
+  private renderContainers(config: ContainerDefinition[] | undefined): {[key: string]: any} {
+    return (config) ? { Containers: config.map(container => (this.renderContainer(container))) } : {};
+  }
+
+  private renderContainer(container: ContainerDefinition): {[key: string]: any} {
+    return (container) ? {
+      ...(container.containerHostName) ? { ContainerHostname: container.containerHostName } : {},
+      ...(container.image) ? { Image: container.image.bind(this).imageUri } : {},
+      ...(container.mode) ? { Mode: container.mode } : {},
+      ...(container.modelDataUrl) ? { ModelDataUrl: container.modelDataUrl } : {},
+      ...(container.modelPackageName) ? { ModelPackageName: container.modelPackageName } : {},
+    } : {};
+  }
```

Comment:
should this be refactored out to someplace more common?

instead of rendering all of these in place, I think they should use the `bind` pattern where a config object is returned with the properties to be attached in `_renderTask` - similar to what we do with the ECS connector. the `render` pattern used in the experimental modules is not one that we should proliferate as it leads to some pretty gigantic classes and

Reply:
Changed ContainerDefinition to follow the bind pattern